<launch>

  <!-- lateral motion -->

  <node name="controller" pkg="pid" type="controller" output="screen" ns="lat_motion">
     <param name="Kp" value="0.0" />
     <param name="Ki" value="0.0" />
     <param name="Kd" value="0.0" />
     <param name="upper_limit" value="1.0" />
     <param name="lower_limit" value="-1.0" />
     <param name="windup_limit" value="0.5" />
     <param name="max_loop_frequency" value="100.0" />
     <param name="min_loop_frequency" value="100.0" />
  </node>
  <node name="pub_lat_setpoint" pkg="topic_tools" type="transform" args="/rov/cmd_vel /lat_motion/setpoint std_msgs/Float64 'm.linear.x'" />
  <node name="pub_lat_vel" pkg="topic_tools" type="transform" respawn="true" args="/t265/odom/sample /lat_motion/state std_msgs/Float64 'm.twist.twist.linear.y'" />

  <!-- longitudinal motion -->

  <node name="controller" pkg="pid" type="controller" output="screen" ns="long_motion">
     <param name="Kp" value="0.0" />
     <param name="Ki" value="0.0" />
     <param name="Kd" value="0.0" />
     <param name="upper_limit" value="1.0" />
     <param name="lower_limit" value="-1.0" />
     <param name="windup_limit" value="0.5" />
     <param name="max_loop_frequency" value="100.0" />
     <param name="min_loop_frequency" value="100.0" />
  </node>
  <node name="pub_long_setpoint" pkg="topic_tools" type="transform" args="/rov/cmd_vel /long_motion/setpoint std_msgs/Float64 'm.linear.y'" />
  <node name="pub_long_vel" pkg="topic_tools" type="transform" respawn="true" args="/t265/odom/sample /long_motion/state std_msgs/Float64 'm.twist.twist.linear.x * -1'" />

  <!-- vertical motion -->

  <node name="controller" pkg="pid" type="controller" output="screen" ns="vert_motion">
     <param name="Kp" value="0.0" />
     <param name="Ki" value="0.0" />
     <param name="Kd" value="0.0" />
     <param name="upper_limit" value="1.0" />
     <param name="lower_limit" value="-1.0" />
     <param name="windup_limit" value="0.5" />
     <param name="max_loop_frequency" value="100.0" />
     <param name="min_loop_frequency" value="100.0" />
  </node>
  <node name="pub_vert_setpoint" pkg="topic_tools" type="transform" args="/rov/cmd_vel /vert_motion/setpoint std_msgs/Float64 'm.linear.z'" />
  <node name="pub_vert_vel" pkg="topic_tools" type="transform" respawn="true" args="/rov/depth_odom /vert_motion/state std_msgs/Float64 'm.twist.twist.linear.z'" />

  <!-- roll motion -->

  <node name="controller" pkg="pid" type="controller" output="screen" ns="roll_motion">
     <param name="Kp" value="0.0" />
     <param name="Ki" value="0.0" />
     <param name="Kd" value="0.0" />
     <param name="upper_limit" value="1.0" />
     <param name="lower_limit" value="-1.0" />
     <param name="windup_limit" value="0.5" />
     <param name="max_loop_frequency" value="100.0" />
     <param name="min_loop_frequency" value="100.0" />
  </node>
  <node name="pub_roll_setpoint" pkg="topic_tools" type="transform" args="/rov/cmd_vel /roll_motion/setpoint std_msgs/Float64 'm.angular.y'" />
  <node name="pub_roll_vel" pkg="topic_tools" type="transform" respawn="true" args="/t265/odom/sample /roll_motion/state std_msgs/Float64 'm.twist.twist.angular.y'" />

  <!-- yaw motion -->

  <node name="controller" pkg="pid" type="controller" output="screen" ns="yaw_motion">
     <param name="Kp" value="0.0" />
     <param name="Ki" value="0.0" />
     <param name="Kd" value="0.0" />
     <param name="upper_limit" value="1.0" />
     <param name="lower_limit" value="-1.0" />
     <param name="windup_limit" value="0.5" />
     <param name="max_loop_frequency" value="100.0" />
     <param name="min_loop_frequency" value="100.0" />
  </node>
  <node name="pub_yaw_setpoint" pkg="topic_tools" type="transform" args="/rov/cmd_vel /yaw_motion/setpoint std_msgs/Float64 'm.angular.x'" />
  <node name="pub_yaw_vel" pkg="topic_tools" type="transform" respawn="true" args="/t265/odom/sample /yaw_motion/state std_msgs/Float64 'm.twist.twist.angular.x'" />

  <!-- combined publisher -->

  <node name="motion_pids_eff_pub" pkg="motion_pids" type="effort_pub" output="screen"/>

  <!-- start pids off by default -->

  <node pkg="rostopic" type="rostopic" name="roll_off_default" args="pub -1 /roll_motion/pid_enable std_msgs/Bool 'data: false'" output="screen"/>
  <node pkg="rostopic" type="rostopic" name="yaw_off_default" args="pub -1 /yaw_motion/pid_enable std_msgs/Bool 'data: false'" output="screen"/>
  <node pkg="rostopic" type="rostopic" name="lat_off_default" args="pub -1 /lat_motion/pid_enable std_msgs/Bool 'data: false'" output="screen"/>
  <node pkg="rostopic" type="rostopic" name="long_off_default" args="pub -1 /long_motion/pid_enable std_msgs/Bool 'data: false'" output="screen"/>
  <node pkg="rostopic" type="rostopic" name="vert_off_default" args="pub -1 /vert_motion/pid_enable std_msgs/Bool 'data: false'" output="screen"/>

</launch>
